executor/insert_executor.rs

```diff
@@ -0,0 +1,199 @@
+};
+use concept::{
+    error::ConceptWriteError,
+    thing::{object::ObjectAPI, thing_manager::ThingManager},
+    type_::{type_manager::TypeManager, Ordering},
+};
+use encoding::value::value::Value;
+use ir::pattern::constraint::{Constraint, Isa};
+use storage::snapshot::WritableSnapshot;
+
+use crate::{batch::Row, VariablePosition};
+
+// TODO: Move to utils
+macro_rules! filter_variants {
+    ($variant:path : $iterable:expr) => {
+        $iterable.iter.filter_map(|item| if let $variant(inner) = item { Some(inner) } else { None })
+    };
+}
+macro_rules! try_unwrap_as {
+    ($variant:path : $item:expr) => {
+        if let $variant(inner) = $item {
+            Some(inner)
+        } else {
+            None
+        }
+    };
+}
+
+struct ExecutionConcepts<'a, 'row> {
+    input: &'a Row<'row>,
+    type_constants: &'a Vec<answer::Type>,
+    value_constants: &'a Vec<Value<'static>>,
+    created_things: &'a mut Vec<answer::Thing<'static>>,
+}
+
+impl<'a, 'row> ExecutionConcepts<'a, 'row> {
+    fn new(
+        input: &'a Row<'row>,
+        type_constants: &'a Vec<answer::Type>,
+        value_constants: &'a Vec<Value<'static>>,
+        created_things: &'a mut Vec<answer::Thing<'static>>,
+    ) -> ExecutionConcepts<'a, 'row> {
+        // TODO: Should we keep an attribute cache to avoid re-creating attributes with the same value?
+        ExecutionConcepts { input, type_constants, value_constants, created_things }
+    }
+
+    fn get_input(&self, position: &usize) -> &VariableValue<'static> {
+        todo!()
+    }
+
+    fn get_value(&self, at: &ValueSource) -> &Value<'static> {
+        match at {
+            ValueSource::ValueConstant(index) => self.value_constants.get(*index).unwrap(),
+            ValueSource::Input(position) => match self.get_input(position) {
+                VariableValue::Value(value) => value,
+                _ => unreachable!(),
+            },
+        }
+    }
+
+    fn get_type(&self, at: &TypeSource) -> &answer::Type {
+        match at {
+            TypeSource::TypeConstant(index) => self.type_constants.get(*index).unwrap(),
+            TypeSource::Input(position) => match self.get_input(position) {
+                VariableValue::Type(type_) => type_,
+                _ => unreachable!(),
+            },
+        }
+    }
+
+    fn get_thing(&self, at: &ThingSource) -> &answer::Thing<'static> {
+        match at {
+            ThingSource::Inserted(index) => self.created_things.get(*index).unwrap(),
+            ThingSource::Input(position) => match self.get_input(position) {
+                VariableValue::Thing(thing) => thing,
+                _ => unreachable!(),
+            },
+        }
+    }
+
+    fn todo__get_thing_list(&self, at: ()) -> &VariableValue<'static> {
+        todo!("???")
+    }
+}
+
+pub struct InsertExecutor {
+    plan: InsertPlan,
+    reused_created_things: Vec<answer::Thing<'static>>, // internal mutability is cleaner
```

Comment:
Re-used vector for pushing created concepts. The order follows that of instructions in the insertion plan